interface/ceed-vector.c

```diff
@@ -326,6 +352,15 @@ int CeedVectorTakeArray(CeedVector vec, CeedMemType mem_type,
                      "has read access");
   // LCOV_EXCL_STOP
 
+  bool has_valid_array = true;
+  ierr = CeedVectorHasValidArray(vec, &has_valid_array); CeedChk(ierr);
+  if (!has_valid_array)
+    // LCOV_EXCL_START
+    return CeedError(vec->ceed, CEED_ERROR_BACKEND,
+                     "CeedVector has no valid data to take, "
+                     "must set data with CeedVectorSetValue or CeedVectorSetArray");
```

Comment:
Is this a change in the new rule for `TakeArray`?  The error messages in `CeedVectorTakeArray_Cuda` still say that you must call `CeedVectorSetArray` with `CEED_USE_POINTER` in order to call `TakeArray`, while here it says `CeedVectorSetValue` would be fine, as well (though `SetValue` could potentially allocate its own memory).  Maybe this one is supposed to have a different error message than the other functions using `HasValidArray`?

Reply:
Two things need to happen

1) the user must set the pointer with SetArray

2) the data must be valid

HasValidArray only tells us if the data is valid, condition 2. Calling SetValue on a user provided array can fix the invalid data issue. Not sure if there is a better way to put that in the two error messages.

Reply:
Perhaps a interface level check HasBorrowedArrayOfType

Reply:
Right, I meant that for this function (and this function only) the rule is not just that the data has to be valid, but that it had to have been set with `SetArray` + `CEED_USE_POINTER` combo, so using `SetValue` for validity purposes not allowed.  Or more specifically, I was noting that the error messages for the interface level function and the Cuda backend version don't match (and the backend version is the one that follows what I thought was the desired rule).

So then the `HasBorrowedArrayOfType` would specifically check for borrowed data?

Reply:
>  but that it had to have been set with `SetArray` + `CEED_USE_POINTER` combo, so using `SetValue` for validity purposes not allowed

I think it would be ok for a user to call `SetValue` after `SetArray` to restore validity of data at a location, which is what I was trying to say. Let me see if I can clean up this error message.